frontend/src/components/LogVisualizer.jsx

```diff
@@ -73,112 +79,123 @@ const buildLineElems = (
-            logDict[logItem.key] = logItem.value;
-          });
-          if (logDict[xAxisKey] == null || logDict[logKey] == null) {
-            return;
-          }
-          if (dataDict[logDict[xAxisKey]] == null) {
-            dataDict[logDict[xAxisKey]] = { [xAxisKey]: logDict[xAxisKey] };
-          }
-          dataDict[logDict[xAxisKey]][line2dataKey(line, axisName)] = logDict[logKey];
-        });
-      });
-    });
-  });
-  const data = Object.keys(dataDict).map((key) => (dataDict[key]));
+class LogVisualizer extends React.Component {
+  constructor(props) {
+    super(props);
 
-  const lineElems = [
-    ...buildLineElems(selectedResults, selectedLogKeys.yLeftAxis, 'yLeftAxis', results, projectConfig, logKeys),
-    ...buildLineElems(selectedResults, selectedLogKeys.yRightAxis, 'yRightAxis', results, projectConfig, logKeys)
-  ];
+    this.chart = null;
 
-  const { chartSize, isResultNameAlignRight } = globalConfig;
+    this.chartRef = this.chartRef.bind(this);
+    this.handleClickDownloadJSON = this.handleClickDownloadJSON.bind(this);
+    this.handleClickDownloadPNG = this.handleClickDownloadPNG.bind(this);
+  }
 
-  return (
-    <div className="log-visualizer-root">
-      <ResponsiveContainer
-        width={chartSize.width}
-        height={chartSize.height}
-        aspect={chartSize.aspect}
-      >
-        <LineChart data={data}>
-          <XAxis
-            type="number"
-            dataKey={xAxisKey}
-            scale={xAxis.scale}
-            domain={getDomain(xAxis)}
-            allowDataOverflow
-          />
-          <YAxis
-            yAxisId="yLeftAxis"
-            orientation="left"
-            scale={yLeftAxis.scale}
-            domain={getDomain(yLeftAxis)}
-            tickFormatter={formatLogValue()}
-            allowDataOverflow
-          />
-          <YAxis
-            yAxisId="yRightAxis"
-            orientation="right"
-            scale={yRightAxis.scale}
-            domain={getDomain(yRightAxis)}
-            tickFormatter={formatLogValue()}
-            allowDataOverflow
-          />
-          <CartesianGrid strokeDasharray="3 3" />
-          <Tooltip
-            content={
-              <LogVisualizerTooltip
-                project={project}
-                results={results}
-                xAxisKey={xAxisKey}
-                isResultNameAlignRight={isResultNameAlignRight}
-              />
+  chartRef(element) {
+    this.chart = element;
+  }
+
+  handleClickDownloadJSON() {
+    const { project, results, stats, projectConfig } = this.props;
+    const data = getLogData(results, stats, projectConfig);
+    const exportName = getUrlSafeProjectNameFull(project);
+    downloadObjectAsJson(data, exportName);
+  }
+
+  handleClickDownloadPNG() {
+    const { project } = this.props;
+    const exportName = getUrlSafeProjectNameFull(project);
+    // eslint-disable-next-line react/no-find-dom-node
+    downloadChartAsPng(ReactDOM.findDOMNode(this.chart), exportName);
```

Comment:
I think I need `ReactDOM.findDOMNode` because `this.chart` is not an ordinal DOM node.
`CategoricalChartWrapper {props: {…}, context: {…}, refs: {…}, updater: {…}, handleLegendBBoxUpdate: f, …}`

I read the document of recharts again but I couldn't find any officially supported ways to get DOM node of recharts component.